src/Hid.Shared/Hid.Helpers.cs

```diff
@@ -0,0 +1,137 @@
+                case NTStatus.HIDP_STATUS_SUCCESS:
+                    return hidCaps;
+
+                case NTStatus.HIDP_STATUS_INVALID_PREPARSED_DATA:
+                    throw new ArgumentException("The specified preparsed data is invalid.", nameof(preparsedData));
+
+                default:
+                    result.ThrowOnError();
+                    return default(HidpCaps);
+            }
+        }
+
+        public static bool HidD_GetManufacturerString(SafeFileHandle hidDeviceObject, out string result)
+        {
+            return GrowBuffer(sb => HidD_GetManufacturerString(hidDeviceObject, sb, sb.Capacity), out result);
+        }
+
+        public static bool HidD_GetProductString(SafeFileHandle hidDeviceObject, out string result)
+        {
+            return GrowBuffer(sb => HidD_GetProductString(hidDeviceObject, sb, sb.Capacity), out result);
+        }
+
+        public static bool HidD_GetSerialNumberString(SafeFileHandle hidDeviceObject, out string result)
+        {
+            return GrowBuffer(sb => HidD_GetSerialNumberString(hidDeviceObject, sb, sb.Capacity), out result);
+        }
+
+        public static string HidD_GetManufacturerString(SafeFileHandle hidDeviceObject)
+        {
+            string result;
+            if (!HidD_GetManufacturerString(hidDeviceObject, out result))
+            {
+                throw new Win32Exception();
+            }
+
+            return result;
+        }
+
+        public static string HidD_GetProductString(SafeFileHandle hidDeviceObject)
+        {
+            string result;
+            if (!HidD_GetProductString(hidDeviceObject, out result))
+            {
+                throw new Win32Exception();
+            }
+
+            return result;
+        }
+
+        public static string HidD_GetSerialNumberString(SafeFileHandle hidDeviceObject)
+        {
+            string result;
+            if (!HidD_GetSerialNumberString(hidDeviceObject, out result))
+            {
+                throw new Win32Exception();
+            }
+
+            return result;
+        }
+
+        private static bool GrowBuffer(Func<StringBuilder, bool> nativeMethod, out string result)
+        {
+            // USB Hid maximum size is 126 wide chars + '\0' = 254 bytes, allocating 256 bytes we should never grow
+            // until another HID standard decide otherwise.
+            var stringBuilder = new StringBuilder(256);
+            while (true)
+            {
+                if (nativeMethod(stringBuilder))
+                {
+                    result = stringBuilder.ToString();
+                    return true;
+                }
+
+                var errorCode = (Win32ErrorCode)Marshal.GetLastWin32Error();
+                if (errorCode != Win32ErrorCode.ERROR_INVALID_USER_BUFFER)
+                {
+                    result = null;
+                    return false;
+                }
+
+                stringBuilder.Capacity = stringBuilder.Capacity * 2;
```

Comment:
Should you cap this somehow so that you don't loop until you throw OutOfMemoryException? Is there some reasonable cut off beyond which you simply return the error code?